In [1]:
%matplotlib inline 
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
import random
from sklearn.cross_validation import KFold, cross_val_score, train_test_split
from sklearn import metrics
import statsmodels.api as sm
import scipy.stats as stats 
from sklearn.linear_model import LogisticRegression

from datetime import datetime
random.seed(datetime.now())
warnings.filterwarnings('ignore')

# Make plots larger
plt.rcParams['figure.figsize'] = (10, 6)

#Recognize the gender as male or female based on voice sample data
#label 1 indicates male and label 0 indicates female
df = pd.read_csv('/Users/neelambabel/ADS Folder/Assignment 2/voice-data.csv')
df.head()



/Users/neelambabel/anaconda/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/Users/neelambabel/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...    centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...    0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...    0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...    0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...    0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...    0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [8]:
df.tail()

meanfreq        sd    median       Q25       Q75       IQR      skew  \
3163  0.131884  0.084734  0.153707  0.049285  0.201144  0.151859  1.762129   
3164  0.116221  0.089221  0.076758  0.042718  0.204911  0.162193  0.693730   
3165  0.142056  0.095798  0.183731  0.033424  0.224360  0.190936  1.876502   
3166  0.143659  0.090628  0.184976  0.043508  0.219943  0.176435  1.591065   
3167  0.165509  0.092884  0.183044  0.070072  0.250827  0.180756  1.705029   

          kurt    sp.ent       sfm   ...    centroid   meanfun    minfun  \
3163  6.630383  0.962934  0.763182   ...    0.131884  0.182790  0.083770   
3164  2.503954  0.960716  0.709570   ...    0.116221  0.188980  0.034409   
3165  6.604509  0.946854  0.654196   ...    0.142056  0.209918  0.039506   
3166  5.388298  0.950436  0.675470   ...    0.143659  0.172375  0.034483   
3167  5.769115  0.938829  0.601529   ...    0.165509  0.185607  0.062257   

        maxfun   meandom    mindom    maxdom   dfrange   modindx   label  
3163  0.262295  0.832899  0.007812  4.210938  4.203125  0.161929  female  
3164  0.275862  0.909856  0.039062  3.679688  3.640625  0.277897  female  
3165  0.275862  0.494271  0.007812  2.937500  2.929688  0.194759  female  
3166  0.250000  0.791360  0.007812  3.593750  3.585938  0.311002  female  
3167  0.271186  0.227022  0.007812  0.554688  0.546875  0.350000  female  

[5 rows x 21 columns]

In [9]:
df.drop(df.index[3167])

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0     0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1     0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2     0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3     0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4     0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5     0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6     0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7     0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8     0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9     0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   
10    0.157021  0.071943  0.168160  0.101430  0.216740  0.115310   0.979442   
11    0.138551  0.077054  0.127527  0.087314  0.202739  0.115426   1.626770   
12    0.137343  0.080877  0.124263  0.083145  0.209227  0.126082   1.378728   
13    0.181225  0.060042  0.190953  0.128839  0.229532  0.100693   1.369430   
14    0.183115  0.066982  0.191233  0.129149  0.240152  0.111004   3.568104   
15    0.174272  0.069411  0.190874  0.115602  0.228279  0.112677   4.485038   
16    0.190846  0.065790  0.207951  0.132280  0.244357  0.112076   1.562304   
17    0.171247  0.074872  0.152807  0.122391  0.243617  0.121227   3.207170   
18    0.168346  0.074121  0.145618  0.115756  0.239824  0.124068   2.704335   
19    0.173631  0.073352  0.153569  0.123680  0.244234  0.120554   2.804975   
20    0.172754  0.076903  0.177736  0.120070  0.245368  0.125298   2.967765   
21    0.181015  0.074369  0.169299  0.128673  0.254175  0.125502   2.587325   
22    0.163536  0.072449  0.145543  0.113930  0.227449  0.113519   3.587650   
23    0.170213  0.075105  0.146053  0.123989  0.250126  0.126137   2.816793   
24    0.160422  0.076615  0.144824  0.120924  0.237244  0.116319   6.253208   
25    0.164700  0.075362  0.147018  0.118698  0.240475  0.121777   4.208608   
26    0.169579  0.075635  0.186468  0.116706  0.238549  0.121843   4.269923   
27    0.169021  0.071778  0.143168  0.125801  0.248315  0.122515   3.079273   
28    0.167340  0.072841  0.141739  0.122174  0.240000  0.117826   2.192126   
29    0.180528  0.070867  0.142385  0.129541  0.252477  0.122936   2.799969   
...        ...       ...       ...       ...       ...       ...        ...   
3137  0.134626  0.084403  0.154531  0.045594  0.209173  0.163580   1.140497   
3138  0.114477  0.081973  0.090199  0.041095  0.199900  0.158806   1.103680   
3139  0.112769  0.074424  0.094248  0.049183  0.183235  0.134052   0.945953   
3140  0.126439  0.079412  0.127325  0.046889  0.198993  0.152103   1.452173   
3141  0.117350  0.090035  0.109478  0.024017  0.203946  0.179929   2.610623   
3142  0.104793  0.085201  0.077886  0.028388  0.186101  0.157712   2.419127   
3143  0.127633  0.084931  0.158892  0.034531  0.201430  0.166899   1.591174   
3144  0.091250  0.086956  0.048191  0.015193  0.179043  0.163851   3.089787   
3145  0.082404  0.085136  0.035114  0.016920  0.152827  0.135906   2.570944   
3146  0.124695  0.080989  0.131882  0.042033  0.197268  0.155234   1.970756   
3147  0.131566  0.084354  0.131889  0.053093  0.196147  0.143055   2.243370   
3148  0.108888  0.092021  0.070063  0.022520  0.201180  0.178660   2.235435   
3149  0.090445  0.079045  0.059358  0.020893  0.167727  0.146834   2.187161   
3150  0.137507  0.091521  0.161298  0.043547  0.221260  0.177713   1.119608   
3151  0.113148  0.090335  0.084335  0.026622  0.198830  0.172207   2.258273   
3152  0.149731  0.082852  0.180932  0.060212  0.219788  0.159576   1.240037   
3153  0.189614  0.035933  0.194116  0.168434  0.205289  0.036855   2.724415   
3154  0.200097  0.045533  0.203796  0.176581  0.232133  0.055552   1.160197   
3155  0.178573  0.046679  0.164388  0.149309  0.204601  

In [10]:
df.isnull().values.any().sum()

0

In [11]:
df = pd.read_csv('/Users/neelambabel/ADS Folder/Assignment 2/voice-data.csv')
# mark zero values as missing or NaN
df["meanfreq"] = df["meanfreq"].replace(0, np.NaN)
# fill missing values with mean column values
df.fillna(df.mean(), inplace=True)
# count the number of NaN values in each column
print(df.isnull().sum())

meanfreq    0
sd          0
median      0
Q25         0
Q75         0
IQR         0
skew        0
kurt        0
sp.ent      0
sfm         0
mode        0
centroid    0
meanfun     0
minfun      0
maxfun      0
meandom     0
mindom      0
maxdom      0
dfrange     0
modindx     0
label       0
dtype: int64


In [12]:
df.isnull().values.any().sum()

0

In [13]:
predictor_names = ['meanfreq',
 'sd',
 'median',
 'Q25',
 'Q75',
 'IQR',
 'skew',
 'kurt',
 'sp.ent',
 'sfm',
 'mode',
 'centroid',
 'meanfun',
 'minfun',
 'maxfun',
 'meandom',
 'mindom',
 'maxdom',
 'dfrange',
 'modindx']

In [14]:
predictor_names

['meanfreq',
 'sd',
 'median',
 'Q25',
 'Q75',
 'IQR',
 'skew',
 'kurt',
 'sp.ent',
 'sfm',
 'mode',
 'centroid',
 'meanfun',
 'minfun',
 'maxfun',
 'meandom',
 'mindom',
 'maxdom',
 'dfrange',
 'modindx']

In [15]:
target= ['label']

In [16]:
target

['label']

In [17]:
#setting predictor fields
x = df[predictor_names]
print(x.shape,'X')
#setting target
y = df[target]
print(y.shape,'Y')

(3168, 20) X
(3168, 1) Y


In [18]:
x[:11]

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0   0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1   0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2   0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3   0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4   0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5   0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6   0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7   0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8   0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9   0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   
10  0.157021  0.071943  0.168160  0.101430  0.216740  0.115310   0.979442   

           kurt    sp.ent       sfm      mode  centroid   meanfun    minfun  \
0    274.402906  0.893369  0.491918  0.000000  0.059781  0.084279  0.015702   
1    634.613855  0.892193  0.513724  0.000000  0.066009  0.107937  0.015826   
2   1024.927705  0.846389  0.478905  0.000000  0.077316  0.098706  0.015656   
3      4.177296  0.963322  0.727232  0.083878  0.151228  0.088965  0.017798   
4      4.333713  0.971955  0.783568  0.104261  0.135120  0.106398  0.016931   
5      8.308895  0.963181  0.738307  0.112555  0.132786  0.110132  0.017112   
6      5.987498  0.967573  0.762638  0.086197  0.150762  0.105945  0.026230   
7      4.766611  0.959255  0.719858  0.128324  0.160514  0.093052  0.017758   
8      4.070284  0.970723  0.770992  0.219103  0.142239  0.096729  0.017957   
9      4.787310  0.975246  0.804505  0.011699  0.134329  0.105881  0.019300   
10     3.974223  0.965249  0.733693  0.096358  0.157021  0.088894  0.022069   

      maxfun   meandom    mindom    maxdom   dfrange   modindx  
0   0.275862  0.007812  0.007812  0.007812  0.000000  0.000000  
1   0.250000  0.009014  0.007812  0.054688  0.046875  0.052632  
2   0.271186  0.007990  0.007812  0.015625  0.007812  0.046512  
3   0.250000  0.201497  0.007812  0.562500  0.554688  0.247119  
4   0.266667  0.712812  0.007812  5.484375  5.476562  0.208274  
5   0.253968  0.298222  0.007812  2.726562  2.718750  0.125160  
6   0.266667  0.479620  0.007812  5.312500  5.304688  0.123992  
7   0.144144  0.301339  0.007812  0.539062  0.531250  0.283937  
8   0.250000  0.336476  0.007812  2.164062  2.156250  0.148272  
9   0.262295  0.340365  0.015625  4.695312  4.679688  0.089920  
10  0.117647  0.460227  0.007812  2.812500  2.804688  0.200000

In [19]:
y[:11]

label
0   male
1   male
2   male
3   male
4   male
5   male
6   male
7   male
8   male
9   male
10  male

In [20]:
type(x)

pandas.core.frame.DataFrame

In [21]:
type(y)

pandas.core.frame.DataFrame

In [22]:
# setting Logistic regression classifier
# scikit-learn includes the intercept.

lr = LogisticRegression(C=1e9)

#dividing data to have a training and a testing set
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= .2, random_state=0)

# Logistic regression metrics
LRm = lr.fit(x_train, y_train)



LRm.predict_proba(x_test)  # The returned estimates for all classes are ordered by the label of classes.


array([[  9.81823861e-01,   1.81761386e-02],
       [  1.33887597e-02,   9.86611240e-01],
       [  9.99998815e-01,   1.18540797e-06],
       ..., 
       [  5.21119660e-02,   9.47888034e-01],
       [  9.89170264e-01,   1.08297360e-02],
       [  9.99417616e-01,   5.82383628e-04]])

In [23]:
LRm.coef_

array([[  5.15874075e+00,   1.20526806e+01,  -3.98469679e+00,
         -1.74053252e+01,   1.92052565e+01,   3.66105832e+01,
          1.76130734e-01,  -8.92526560e-03,   3.18689580e+01,
         -9.25096449e+00,   1.99477609e+00,   5.15874075e+00,
         -1.61574787e+02,   3.17413787e+01,  -2.66508177e+00,
         -2.95257870e-02,  -1.30267371e+00,  -6.54992301e-01,
          6.47680664e-01,  -2.83120591e+00]])

In [24]:
LRm.intercept_

array([-9.6494162])

In [25]:
LRm.decision_function(x_test) # Predict confidence scores for samples.

array([ -3.98930226e+00,   4.29986055e+00,  -1.36454224e+01,
        -9.26673867e+00,   6.46770239e+00,   4.27442900e+00,
        -7.15298808e+00,  -4.17513963e+00,   5.21715306e+00,
        -7.10130849e+00,   5.30764097e+00,  -7.86047469e+00,
        -1.40313899e+00,   4.61134353e+00,  -4.72341390e-01,
        -8.12404448e+00,  -4.13604083e+00,   4.09663125e+00,
         1.71378565e+00,  -8.25088403e+00,  -1.25210839e+01,
         7.39503474e+00,   5.15677090e+00,  -8.71005852e+00,
        -4.55565951e+00,   2.39857383e+00,  -7.10906859e+00,
        -8.47242575e+00,  -3.81668506e+00,   8.49803224e+00,
        -3.06992578e+00,  -7.86066715e+00,   6.51184047e+00,
         1.05174191e+00,   3.42208688e+00,  -1.06812046e+01,
         7.31000695e+00,  -5.99691715e+00,   5.10749055e+00,
         5.39669155e+00,   5.57447930e+00,  -6.89917534e+00,
        -9.20711624e+00,   6.60729946e+00,  -4.86579726e+00,
         1.34795771e+01,   7.53518876e-01,  -7.14904200e+00,
         2.98932873e+00,

In [26]:
LRm.predict(x_test)


array(['female', 'male', 'female', 'female', 'male', 'male', 'female',
       'female', 'male', 'female', 'male', 'female', 'female', 'male',
       'female', 'female', 'female', 'male', 'male', 'female', 'female',
       'male', 'male', 'female', 'female', 'male', 'female', 'female',
       'female', 'male', 'female', 'female', 'male', 'male', 'male',
       'female', 'male', 'female', 'male', 'male', 'male', 'female',
       'female', 'male', 'female', 'male', 'male', 'female', 'male',
       'female', 'male', 'male', 'male', 'female', 'male', 'male',
       'female', 'female', 'female', 'male', 'male', 'male', 'female',
       'male', 'male', 'female', 'female', 'male', 'female', 'female',
       'male', 'male', 'male', 'male', 'male', 'female', 'male', 'male',
       'male', 'female', 'female', 'male', 'male', 'female', 'male',
       'male', 'female', 'female', 'male', 'male', 'female', 'male',
       'female', 'male', 'female', 'female', 'male', 'male', 'male',
       'female', '

In [27]:
y_pred = LRm.predict(x_test)
print("Accuracy score using Logistic regression is %s" %metrics.accuracy_score(y_test, y_pred))

Accuracy score using Logistic regression is 0.977917981073


In [28]:
k=10
acc=[]
for i in range(k):
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= 0.2, random_state=i)
    # Logistic regression metrics
    LRm = lr.fit(x_train, y_train)
    a=metrics.accuracy_score(y_test, LRm.predict(x_test))
    acc.append(a)
    
print (acc)
print (np.mean(acc))
print (np.std(acc))

[0.97791798107255523, 0.96845425867507884, 0.97160883280757093, 0.97318611987381698, 0.97003154574132489, 0.97003154574132489, 0.97003154574132489, 0.96687697160883279, 0.97318611987381698, 0.98107255520504733]
0.972239747634
0.00412514095603


Summary of the stats:

* Is the relationship significant?
  Yes, the relationship is significant. We can see that the coefficient values,i.e. the slope, changes significantly,   implying the variations in the response variable (gender) w.r.t the predictor variables (meanfun, sp.ent etc) 

* Are any model assumptions violated? 
  The model assumes that the response variable should be categorical, hence there are no violations.

* Cross-validate the model. How well did it do?
  We split the training and test data into 80% and 20% resp. We cross-validated the consistency in accuracy score 10     times with different sets of training data and test data, and we've achieved an accuracy score of 97.2% with a         standard deviation of 4%.
